### ISSUE TITLE
logstash formatter - ElasticSearch v2 does not support field names with different types

### ISSUE THREAD

**Issue by lonnc** [Sirupsen/logrus#300]
See https://www.elastic.co/blog/great-mapping-refactoring#conflicting-mappings

If a field name previously seen by ElasticSearch v2 has a type of say 'string' caused by a log statement like:
```
log.WithFields(logrus.Fields{
                "animal": "walrus",
        }).Info("A group of walrus emerges from the ocean")
```
and later pass in a map:
```
log.WithFields(logrus.Fields{
                "animal": map[string]interface{}{
                       "id": 1,
                       "name": "dog",
                       "legs": 4,
                 },
        }).Info("A dog")
```
then when passed to logstash (using the json filter) for import into ElasticSearch v2 you get a logstash error like:

"status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"failed to parse [animal]", "caused_by"=>{"type"=>"illegal_argument_exception", "reason"=>"unknown property [id]"}

No record is created in elastic search.

Simple work around as suggested by https://www.elastic.co/blog/great-mapping-refactoring#conflicting-mappings is to rename any conflicting fields, but this requires changes to the logging code in the application once conflicts have been identified.

If possible would be great if the logstash formatter could handle such cases, or best practise given to prevent such situations.

**Comment by aybabtme** [comment]
@lonnc thanks for reporting this. 

I think those cases are application bugs. If someone is using a logstash formatter but does not respect the peculiarities of logstash, it's my belief that this application should be fixed. I don't think it's practical for `logrus` to be checking that entry field types don't change over time.  

Maybe if Go had a stronger type system, it would be possible to enforce at compile time when using the logstash formatter. But otherwise, trying to help people write correct code in this case means we'd have to break their current usage of the formatter with runtime checks: means we'd give bad surprises to people.

In short: because I think those cases are application bugs, I'll be closing this issue.
